### PR TITLE
NIFI-12513 Added UriUtils class which allows the construction of a valid java.net.URI with a single string even though there maybe illegal characters in the path, query and/or fragment sections(s) of the URI.

### PR DESCRIPTION
**Pull request by dan-s1**
…

# Summary

[NIFI-12513](https://issues.apache.org/jira/browse/NIFI-12513)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/UriUtils.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+     *  </ul>
+     *  On the other hand, java.net.URI's seven argument constructor provides these quoting capabilities. In order
+     *  to take advantage of this constructor, this method parses the given string into the arguments needed
+     *  thereby allowing for instantiating a java.net.URI with the quoting of all illegal characters.
+     * @param uri String representing a URI.
+     * @return Instance of java.net.URI
+     */
+    public static URI create(String uri) throws URISyntaxException {
+        final Matcher matcher = URI_PATTERN.matcher(uri);
+        if(matcher.matches()) {
```
Comment (exceptionfactory) [inline nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/UriUtils.java:64]: Spacing:
```suggestion
        if (matcher.matches()) {
```

File: nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/UriUtils.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+     *   <ul>
+     *      <li>They do not provide quoting in the path section for any character not in the unreserved, punct, escaped, or other categories,
+     *          and not equal to the slash character ('/') or the commercial-at character ('{@literal @}').</li>
+     *      <li>They do not provide quoting for any illegal characters found in the query and fragment sections.</li>
+     *  </ul>
+     *  On the other hand, java.net.URI's seven argument constructor provides these quoting capabilities. In order
+     *  to take advantage of this constructor, this method parses the given string into the arguments needed
+     *  thereby allowing for instantiating a java.net.URI with the quoting of all illegal characters.
+     * @param uri String representing a URI.
+     * @return Instance of java.net.URI
```
Comment (exceptionfactory) [inline nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/UriUtils.java:60]: A new line should be added along the lines of `@throws URISyntaxException Thrown on parsing failures`

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for researching the options and putting together this solution @dan-s1. The selected implementation from Spring Framework UriComponentsBuilder looks good. I recommend adding a basic unit test class for `UriUtils`. It does not need to be elaborate, but at least one method that succeeds, and one method that throws an exception would be helpful.

**Comment by dan-s1** [comment]
@exceptionfactory 
The failure on `ci-workflow / Windows Zulu JDK 21 FR ` below does not seem related to my changes
```
 Failures: 
Error:    TestFTP.testListFtpHostPortVariablesFileFound:317 expected: <1> but was: <0>
Error:  Tests run: 1786, Failures: 1, Errors: 0, Skipped: 129
Error:  Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.1.2:test (default-test) on project nifi-standard-processors: There are test failures.
```

**Review by exceptionfactory — APPROVED** [review]
Thanks for adding the unit test @dan-s1, the latest version looks good! +1 merging

**Pull request closed without merging**